p-ata/src/processor.rs

```diff
@@ -0,0 +1,982 @@
+///
+/// ## Returns
+///
+/// `(address, bump)` - The canonical PDA address and its bump seed
+#[inline(always)]
+pub(crate) fn derive_canonical_ata_pda(
+    wallet: &Pubkey,
+    token_program: &Pubkey,
+    mint: &Pubkey,
+    program_id: &Pubkey,
+) -> (Pubkey, u8) {
+    find_program_address(
+        &[wallet.as_ref(), token_program.as_ref(), mint.as_ref()],
+        program_id,
+    )
+}
+
+/// Check if the given program ID is SPL Token (not Token-2022)
+#[inline(always)]
+pub(crate) fn is_spl_token_program(program_id: &Pubkey) -> bool {
+    const SPL_TOKEN_PROGRAM_ID: Pubkey =
+        pinocchio_pubkey::pubkey!("TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA");
+    *program_id == SPL_TOKEN_PROGRAM_ID
+}
+
+/// Check if account data represents an initialized token account.
+/// Mimics p-token's is_initialized_account check.
+///
+/// Panics if account_data.len() < 109.
+#[inline(always)]
+pub(crate) fn is_initialized_account(account_data: &[u8]) -> bool {
+    // Token account state is at offset 108 (after mint, owner, amount, delegate fields)
+    // State: 0 = Uninitialized, 1 = Initialized, 2 = Frozen
+    account_data[108] != 0
+}
+
+/// Validate that account data represents a valid token account.
+/// Replicates Token-2022's GenericTokenAccount::valid_account_data checks.
+#[inline(always)]
+pub(crate) fn valid_token_account_data(account_data: &[u8]) -> bool {
+    // Regular Token account: exact length match and initialized
+    if account_data.len() == TokenAccount::LEN {
+        // SAFETY: TokenAccount::LEN is compile-ensured to be == 165
+        return is_initialized_account(account_data);
+    }
+
+    // Token-2022's GenericTokenAccount::valid_account_data assumes Multisig
+    // if account_data length is Multisig::LEN. Collisions are prevented by
+    // adding a byte if a token account happens to have the same length as
+    // Multisig::LEN.
+    if account_data.len() > TokenAccount::LEN {
+        if account_data.len() == Multisig::LEN {
+            return false;
+        }
+        // SAFETY: TokenAccount::LEN is compile-ensured to be == 165, and in
+        // this branch account_data.len > TokenAccount::LEN
+        if is_initialized_account(account_data) {
+            return account_data[TokenAccount::LEN] == ACCOUNTTYPE_ACCOUNT;
+        }
+    }
+
+    false
+}
+
+/// Get mint reference from account info
+#[inline(always)]
+pub(crate) fn get_decimals_from_mint(account: &AccountInfo) -> Result<u8, ProgramError> {
+    let mint_data_slice = account.try_borrow_data()?;
+    if mint_data_slice.len() < MINT_BASE_SIZE {
+        log!(
+            "Error: Mint account data too small. Expected at least {} bytes, found {} bytes",
+            MINT_BASE_SIZE,
+            mint_data_slice.len()
+        );
+        return Err(ProgramError::InvalidAccountData);
+    }
+    // SAFETY: We've validated the account length above
+    let mint = unsafe { &*(mint_data_slice.as_ptr() as *const Mint) };
```

Comment:
Probably can be replaced by:
```suggestion
    let mint = unsafe { spl_token_interface::state::load_unchecked::<Mint>(&mint_data_slice) };
```